### PR TITLE
CORTX-30213: Added Activity framework 

### PR DESCRIPTION
**Pull request by kailashyadav11**
# Problem Statement
- Add an infrastructure for maintaining information about various ongoing activities / tasks in the cluster and view them asynchronously.

# Design
-  For Bug describe the fix here. 
-  [For Feature, Post the link to the solution page on the confluence CORTX Foundation Library 
- https://seagate-systems.atlassian.net/wiki/spaces/PUB/pages/580550715/WIP+CORTX+Task+Framework](https://seagate-systems.atlassian.net/wiki/spaces/PUB/pages/580550715/WIP+CORTX+Task+Framework)

# Coding 
-  Coding conventions are followed and code is consistent [Y/N]: Y
-  Confirm All CODACY errors are resolved [Y/N]: Y

# Testing
- Are test cases updated along with code changes due to Enhancements/Bugs [Y/N]: Y
- Confirm that new test cases are added to regression and sanity plan files and relevant feature plan files [Y/N]:
- Confirm that Test Cases are added for new features added [Y/N]:  Y
- Confirm Test Cases cover Happy Path, Non-Happy Path and Scalability [Y/N]: Y
- Confirm Testing was performed with installed RPM/K8s deployment [Y/N]:  Y

# Review Checklist 
####  Before posting the PR please ensure:
- [x] PR is self reviewed
- [ ] Is there a change in filename/package/module or signature [Y/N]: 
- [ ] If yes for above point, Is a notification sent to all other cortx components [Y/N]
- [x] New package/s added to setup.py?
- [x] Jira is updated
- [x] Check if the description is clear and explained. 
- [x] Check Acceptance Criterion is defined. 
- [ ] All the tests performed should be mentioned before Resolving a JIRA. 
- [x] Verification needs to be done before marked as Closed/Verified.

# KVstore/Confstore feature (changes/fixes) checklist
#### Confirm changes are made for:
- [ ] ConfStore
- [ ] KVStore
- [ ] ConfCli
- [ ] Test cases added for all above 3
- [ ] changes done in all the KVpayloads (ConsulKvPayload, IniKvPayload, KvPayload)

# Documentation
- [ ] Changes done to WIKI / Confluence page

### CODE REVIEW

**Review by ujjwalpl — CHANGES_REQUESTED** [review]

File: py-utils/test/activity_framework/test_activity_store.py [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+        Activity.finish(activity)
+        try:
+            activities = load_data(sample_file)['Activities']
+            activity_data = json.loads(activities.get(activity_id))
+        except Exception as e:
+            rc = 1
+            print(e)
+        self.assertIsNotNone(activity_data.get('finish_time'), "finish_time key is not present")
+
+    def test_activity_store_search_activity(self):
```
Comment (ujjwalpl) [inline py-utils/test/activity_framework/test_activity_store.py:147]: Need to add more cases of search.

**Review by ujjwalpl — CHANGES_REQUESTED** [review]

File: py-utils/setup.py [hunk 2]
```diff
@@ -114,7 +114,7 @@ def get_requirements_files() -> list:
                 'cortx.utils.support_framework',
                 'cortx.utils.manifest',
                 'cortx.utils.audit_log', 'cortx.utils.cortx',
-                'cortx.utils.http', 'cortx.utils.s3',
+                'cortx.utils.http', 'cortx.utils.s3', 'cortx.utils.activity_framework'
```
[the remark this answers is not in the document]
Comment (tanujashinde0405) [reply py-utils/setup.py]: done

**Review by ujjwalpl — APPROVED** [review]

File: py-utils/src/utils/activity_tracker/__init__.py [hunk 3]
```diff
@@ -0,0 +1,14 @@
+# CORTX-Py-Utils: CORTX Python common library.
+# Copyright (c) 2021 Seagate Technology LLC and/or its Affiliates
```
Comment (selvakumaar5496) [inline py-utils/src/utils/activity_tracker/__init__.py:2]: Copyright (c) 2022

File: py-utils/src/utils/activity_tracker/activity_cli.py [hunk 4]
```diff
@@ -0,0 +1,199 @@
+#!/usr/bin/env python3
+
+# CORTX-Py-Utils: CORTX Python common library.
+# Copyright (c) 2021 Seagate Technology LLC and/or its Affiliates
```
Comment (selvakumaar5496) [inline py-utils/src/utils/activity_tracker/activity_cli.py:4]: Copyright (c) 2022

File: py-utils/src/utils/activity_tracker/__init__.py (hunk 3 above)
Comment (tanujashinde0405) [reply py-utils/src/utils/activity_tracker/__init__.py]: done

**Review by tanujashinde0405 — APPROVED** [review]

File: py-utils/src/utils/activity_tracker/activity_store.py [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+
+from cortx.template import Singleton
+from cortx.utils.activity_tracker.error import ActivityError
+from cortx.utils.kv_store.kv_store import KvStoreFactory
+from cortx.utils.kv_store import KvPayload
+
+class ActivityEntry:
+    """Represents System Activity"""
+
+    def __init__(self, *args, **kwargs):
```
Comment (selvakumaar5496) [inline py-utils/src/utils/activity_tracker/activity_store.py:31]: *args is unused and can be removed

File: py-utils/src/utils/activity_tracker/error.py [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+# GNU Affero General Public License for more details.
+# You should have received a copy of the GNU Affero General Public License
+# along with this program. If not, see <https://www.gnu.org/licenses/>.
+# For any questions about this software or licensing,
+# please email opensource@seagate.com or cortx-questions@seagate.com.
+
+
+class ActivityError(Exception):
+    """Generic Exception with error code and output."""
```
Comment (selvakumaar5496) [inline py-utils/src/utils/activity_tracker/error.py:20]: No need to rewrite this whole file. This class Can be derived from UtilsError

reference , https://github.com/Seagate/cortx-utils/blob/main/py-utils/src/utils/common/errors.py

File: py-utils/src/utils/activity_tracker/activity_store.py [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+            data = json.loads(val[0])
+            matched = True
+            for f in filters:
+                key_spec = re.split("[^a-zA-Z_]", f)
+                if key_spec[0] not in data.keys():
+                    matched = False
+                    break
+                vars()[key_spec[0]] = data[key_spec[0]]
+                _ , r = [i.strip() for i in f.split("==")]
+                r=r.replace("'","")
```
Comment (selvakumaar5496) [inline py-utils/src/utils/activity_tracker/activity_store.py:127]: spaces before and after '='

File: py-utils/src/utils/activity_tracker/activity_store.py [hunk 8]
```diff
@@ -0,0 +1,142 @@
...
+            val = Activity._kv_store.get([activity_id])
+            data = json.loads(val[0])
+            matched = True
+            for f in filters:
+                key_spec = re.split("[^a-zA-Z_]", f)
+                if key_spec[0] not in data.keys():
+                    matched = False
+                    break
+                vars()[key_spec[0]] = data[key_spec[0]]
+                _ , r = [i.strip() for i in f.split("==")]
```
Comment (selvakumaar5496) [inline py-utils/src/utils/activity_tracker/activity_store.py:126]: remove spaces before comma

File: py-utils/src/utils/activity_tracker/activity_cli.py [hunk 9]
```diff
@@ -0,0 +1,199 @@
...
+        s_parser = sub_parser.add_parser('show', help=
+            "Shows the Activity information. Example command:\n"
+            "# activity show '123>1>2>3>222.222'\n\n")
+        s_parser.set_defaults(func=ActivityCli.show)
+        s_parser.add_argument('args', nargs='+', default=[], help='args')
+
+
+def main():
+    # Setup Parser
+    parser = argparse.ArgumentParser(description='Activity CLI',
```
Comment (selvakumaar5496) [inline py-utils/src/utils/activity_tracker/activity_cli.py:165]: Check if cmd framework can be used here. 

reference, https://github.com/Seagate/cortx-rgw-integration/blob/main/src/rgw/setup/rgw_setup

File: py-utils/test/activity_tracker/test_activity_store.py [hunk 10]
```diff
@@ -0,0 +1,167 @@
...
+    """Returns JSON object of the data in JSON file."""
+    data = None
+    try:
+        with open(file,) as f_obj:
+            data = json.load(f_obj)
+    except FileNotFoundError as e:
+        print(e)
+    return data
+
+class TestActivityStore(unittest.TestCase):
```
Comment (selvakumaar5496) [inline py-utils/test/activity_tracker/test_activity_store.py:47]: 1. Printing exceptions is not required
2. Add Unhappy path test cases

**Review by selvakumaar5496 — CHANGES_REQUESTED** [review]

File: py-utils/test/activity_tracker/test_activity_store.py (hunk 10 above)
Comment (tanujashinde0405) [reply py-utils/test/activity_tracker/test_activity_store.py:48]: done

File: py-utils/src/utils/activity_tracker/activity_store.py (hunk 8 above)
Comment (tanujashinde0405) [reply py-utils/src/utils/activity_tracker/activity_store.py]: done

File: py-utils/src/utils/activity_tracker/activity_store.py (hunk 7 above)
Comment (tanujashinde0405) [reply py-utils/src/utils/activity_tracker/activity_store.py]: done

File: py-utils/src/utils/activity_tracker/error.py (hunk 6 above)
Comment (tanujashinde0405) [reply py-utils/src/utils/activity_tracker/error.py:21]: done

File: py-utils/src/utils/activity_tracker/activity_store.py (hunk 5 above)
Comment (tanujashinde0405) [reply py-utils/src/utils/activity_tracker/activity_store.py]: done

File: py-utils/src/utils/activity_tracker/activity_cli.py (hunk 4 above)
Comment (tanujashinde0405) [reply py-utils/src/utils/activity_tracker/activity_cli.py]: done

**Review by selvakumaar5496 — APPROVED** [review]

**Pull request merged**